File: frontend/src/api/auditLog.ts
import { api } from "./client"
import type { PaginatedAuditLog } from "./types"

export const auditLogApi = {
  list: (
    params: {
      entity_type?: string
      entity_id?: string
      user_id?: string
      member_id?: string
      from?: string
      to?: string
      page?: number
      page_size?: number
    } = {},
  ) => {
    const qs = new URLSearchParams()
    if (params.entity_type) qs.set("entity_type", params.entity_type)
    if (params.entity_id) qs.set("entity_id", params.entity_id)
    if (params.user_id) qs.set("user_id", params.user_id)
    if (params.member_id) qs.set("member_id", params.member_id)
    if (params.from) qs.set("from", params.from)
    if (params.to) qs.set("to", params.to)
    if (params.page) qs.set("page", String(params.page))
    if (params.page_size) qs.set("page_size", String(params.page_size))
    const suffix = qs.toString() ? `?${qs.toString()}` : ""
    return api.get<PaginatedAuditLog>(`/audit-log${suffix}`)
  },
}

File: backend/app/api/v1/audit_log.py
import uuid
from datetime import datetime

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.visibility import VisibilityContext, get_visibility_ctx
from app.db.base import get_session
from app.schemas.audit import PaginatedAuditLog
from app.services.audit import AuditLogService

router = APIRouter()


@router.get("/audit-log", response_model=PaginatedAuditLog)
async def list_audit_log(
    entity_type: str | None = Query(None),
    entity_id: uuid.UUID | None = Query(None),
    user_id: uuid.UUID | None = Query(None),
    member_id: uuid.UUID | None = Query(None),
    from_: datetime | None = Query(None, alias="from"),
    to: datetime | None = Query(None),
    page: int = Query(1, ge=1),
    page_size: int = Query(50, ge=1, le=200),
    ctx: VisibilityContext = Depends(get_visibility_ctx),
    session: AsyncSession = Depends(get_session),
) -> PaginatedAuditLog:
    svc = AuditLogService(session)
    return await svc.list_entries(
        ctx,
        entity_type=entity_type,
        entity_id=entity_id,
        user_id=user_id,
        member_id=member_id,
        from_date=from_,
        to_date=to,
        page=page,
        page_size=page_size,
    )
